Flakyness in TestAttachStopOnEntry

TestAttachStopOnEntry fails occasionally as follows:

```
=== RUN   TestAttachStopOnEntry
DAP server listening at: [::]:37341
Connecting to server at: [::]:37341
    server_test.go:294: 
        got &dap.ThreadsResponse{Response:dap.Response{ProtocolMessage:dap.ProtocolMessage{Seq:0, Type:"response"}, RequestSe
q:6, Success:true, Command:"threads", Message:""}, Body:dap.ThreadsResponseBody{Threads:[]dap.Thread{dap.Thread{Id:1, Name:"*
 [Go 1] runtime.GOROOT (Thread 27140)"}, dap.Thread{Id:2, Name:"[Go 2] runtime.gopark"}, dap.Thread{Id:3, Name:"[Go 3] runtim
e.gopark"}, dap.Thread{Id:4, Name:"[Go 4] runtime.gopark"}}}}
        want Seq=0, RequestSeq=6 len(Threads)=1
    server_test.go:297: 
        got &dap.ThreadsResponse{Response:dap.Response{ProtocolMessage:dap.ProtocolMessage{Seq:0, Type:"response"}, RequestSe
q:6, Success:true, Command:"threads", Message:""}, Body:dap.ThreadsResponseBody{Threads:[]dap.Thread{dap.Thread{Id:1, Name:"*
 [Go 1] runtime.GOROOT (Thread 27140)"}, dap.Thread{Id:2, Name:"[Go 2] runtime.gopark"}, dap.Thread{Id:3, Name:"[Go 3] runtim
e.gopark"}, dap.Thread{Id:4, Name:"[Go 4] runtime.gopark"}}}}
        want Id=1, Name="Dummy"
    server_test.go:304: 
        got &dap.ThreadsResponse{Response:dap.Response{ProtocolMessage:dap.ProtocolMessage{Seq:0, Type:"response"}, RequestSe
q:7, Success:true, Command:"threads", Message:""}, Body:dap.ThreadsResponseBody{Threads:[]dap.Thread{dap.Thread{Id:1, Name:"*
 [Go 1] runtime.GOROOT (Thread 27140)"}, dap.Thread{Id:2, Name:"[Go 2] runtime.gopark"}, dap.Thread{Id:3, Name:"[Go 3] runtim
e.gopark"}, dap.Thread{Id:4, Name:"[Go 4] runtime.gopark"}}}}
        want Seq=0, RequestSeq=7 len(Threads)=1
--- FAIL: TestAttachStopOnEntry (0.61s)
panic: interface conversion: dap.Message is *dap.StackTraceResponse, not *dap.ErrorResponse [recovered]
        panic: interface conversion: dap.Message is *dap.StackTraceResponse, not *dap.ErrorResponse

goroutine 50 [running]:
testing.tRunner.func1.1(0x7add40, 0xc0004b8b10)
        /usr/local/go/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc000083b00)
        /usr/local/go/src/testing/testing.go:1075 +0x41a
panic(0x7add40, 0xc0004b8b10)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/go-delve/delve/service/dap/daptest.(*Client).ExpectErrorResponse(0xc000574620, 0xc000083b00, 0x0)
        /home/a/n/delve/service/dap/daptest/client.go:64 +0xf9
github.com/go-delve/delve/service/dap.TestAttachStopOnEntry.func1(0xc000574620, 0x81b6fe, 0x8, 0xc000257240, 0x16, 0xc0005ca120, 0x25, 0xc000607de0, 0x19)
        /home/a/n/delve/service/dap/server_test.go:309 +0x82c
github.com/go-delve/delve/service/dap.runTest(0xc000083b00, 0x81b6fe, 0x8, 0xc00015ff60)
        /home/a/n/delve/service/dap/server_test.go:79 +0x31a
github.com/go-delve/delve/service/dap.TestAttachStopOnEntry(0xc000083b00)
        /home/a/n/delve/service/dap/server_test.go:233 +0x65
testing.tRunner(0xc000083b00, 0x832de8)
        /usr/local/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1168 +0x2b3
```

Reading the tests it looks to me that there's a race condition between the target process and the test logic, if the target process succeeds in initializing the runtime before the test attaches the test will fail (because the main goroutine can be found). Am I missing something?

cc @polinasok

--- a/service/dap/server_test.go
+++ b/service/dap/server_test.go
@@ -1,6 +1,7 @@
 package dap
 
 import (
+	"bufio"
 	"flag"
 	"io"
 	"net"
@@ -232,12 +233,22 @@ func TestAttachStopOnEntry(t *testing.T) {
 	}
 	runTest(t, "loopprog", func(client *daptest.Client, fixture protest.Fixture) {
 		// Start the program to attach to
-		// TODO(polina): do I need to sanity check testBackend and runtime.GOOS?
 		cmd := exec.Command(fixture.Path)
-		cmd.Stdout = os.Stdout
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			t.Fatal(err)
+		}
 		cmd.Stderr = os.Stderr
 		if err := cmd.Start(); err != nil {
 			t.Fatal(err)
+		}
+		// Wait for output.
+		// This will give the target process time to initialize the runtime before we attach,
+		// so we can rely on having goroutines when they are requested on attach.
+		scanOut := bufio.NewScanner(stdout)
+		scanOut.Scan()
+		if scanOut.Text() != "past main" {
+			t.Errorf("expected loopprog.go to output \"past main\"")
 		}
 
 		// 1 >> initialize, << initialize
@@ -247,7 +258,7 @@ func TestAttachStopOnEntry(t *testing.T) {
 			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=1", initResp)
 		}
 
-		// 2 >> attach, << initialized, << launch
+		// 2 >> attach, << initialized, << attach
 		client.AttachRequest(
 			map[string]interface{}{"mode": "local", "processId": cmd.Process.Pid, "stopOnEntry": true})
 		initEvent := client.ExpectInitializedEvent(t)
@@ -290,33 +301,22 @@ func TestAttachStopOnEntry(t *testing.T) {
 		// 6 >> threads, << threads
 		client.ThreadsRequest()
 		tResp := client.ExpectThreadsResponse(t)
-		if tResp.Seq != 0 || tResp.RequestSeq != 6 || len(tResp.Body.Threads) != 1 {
-			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=6 len(Threads)=1", tResp)
-		}
-		if tResp.Body.Threads[0].Id != 1 || tResp.Body.Threads[0].Name != "Dummy" {
-			t.Errorf("\ngot %#v\nwant Id=1, Name=\"Dummy\"", tResp)
+		// Expect main goroutine plus runtime at this point.
+		if tResp.Seq != 0 || tResp.RequestSeq != 6 || len(tResp.Body.Threads) < 2 {
+			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=6 len(Threads)>1", tResp)
 		}
 
 		// 7 >> threads, << threads
 		client.ThreadsRequest()
-		tResp = client.ExpectThreadsResponse(t)
-		if tResp.Seq != 0 || tResp.RequestSeq != 7 || len(tResp.Body.Threads) != 1 {
-			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=7 len(Threads)=1", tResp)
-		}
+		client.ExpectThreadsResponse(t)
 
-		// 8 >> stackTrace, << error
+		// 8 >> stackTrace, << response
 		client.StackTraceRequest(1, 0, 20)
-		stResp := client.ExpectErrorResponse(t)
-		if stResp.Seq != 0 || stResp.RequestSeq != 8 || stResp.Body.Error.Format != "Unable to produce stack trace: unknown goroutine 1" {
-			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=8 Format=\"Unable to produce stack trace: unknown goroutine 1\"", stResp)
-		}
+		client.ExpectStackTraceResponse(t)
 
-		// 9 >> stackTrace, << error
+		// 9 >> stackTrace, << response
 		client.StackTraceRequest(1, 0, 20)
-		stResp = client.ExpectErrorResponse(t)
-		if stResp.Seq != 0 || stResp.RequestSeq != 9 || stResp.Body.Error.Id != 2004 {
-			t.Errorf("\ngot %#v\nwant Seq=0, RequestSeq=9 Id=2004", stResp)
-		}
+		client.ExpectStackTraceResponse(t)
 
 		// 10 >> evaluate, << error
 		client.EvaluateRequest("foo", 0 /*no frame specified*/, "repl")

